### ISSUE TITLE
Issue CodeGen for Crossplane EC2.Volume - VolumeAttachment struct is missing

### ISSUE THREAD

**Issue by haarchri** [aws-controllers-k8s/community#864]
**Describe the bug**
we see with codegen for crossplane EC2.Volume the struct for VolumeAttachment is missing - any ideas why ?

i added in crossplane via custom_types.go the following:
```
// VolumeAttachment
type VolumeAttachment struct {
	// The time stamp when the attachment initiated.
	AttachTime *metav1.Time `json:"attachTime,omitempty"`

	// Indicates whether the EBS volume is deleted on instance termination.
	DeleteOnTermination *bool `json:"deleteOnTermination,omitempty"`

	// The device name.
	Device *string `json:"device,omitempty"`

	// The ID of the instance.
	InstanceID *string `json:"instanceID,omitempty"`

	// The attachment state of the volume.
	State *string `json:"state,omitempty"`

	// The ID of the volume.
	VolumeID *string `json:"volumeID,omitempty"`
}
```

but then i faced the problem / warning in zz_controller.go and zz_conversions.go like
`k8s.io/apimachinery/pkg/apis/meta/v1.Time composite literal uses unkeyed fieldscomposites`

**Environment**
* Crossplane Provider-AWS

**Comment by jaypipes** [comment]
Hi @haarchri, `VolumeAttachment` is not an element of the `Volume` resource. The elements of the `Volume` resource are the fields that appear in the Request and Response shapes from the [CreateVolume API call](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateVolume.html)

**Comment by haarchri** [comment]
Hello @jaypipes  for me it looks that VolumeAttachment is part of `VolumeObservation` which is generated - https://github.com/haarchri/provider-aws/blob/ebs-volume/apis/ec2/v1alpha1/zz_volume.go#L153

```
type VolumeObservation struct {
	// Information about the volume attachments.
	Attachments []*VolumeAttachment `json:"attachments,omitempty"`
```
if the struct for VolumeAttachment is missing the code is not running ... 

any ideas ?

**Comment by jaypipes** [comment]
@haarchri I'm still trying to figure this one out...

One thing I did note, however, is that you need to change this line in your generator config:

https://github.com/haarchri/provider-aws/blob/677ea56479724869f6a33eaaa2358b717309fda3/apis/ec2/v1alpha1/generator-config.yaml#L60

to be:

```yaml
   - CreateVolumeRequest.DryRun
```

The reason is because the "shape" in the EC2 API model is called `CreateVolumeRequest`:

https://github.com/aws/aws-sdk-go/blob/287bf610b41fa8b86e0db1904a103db7ec3ec37f/models/apis/ec2/2016-11-15/api-2.json#L9301

I have a meeting this afternoon with some EC2 devs and I'll see if I can figure out what is going on with your `VolumeAttachment` struct generation. Thanks for your patience!

**Comment by jaypipes** [comment]
@haarchri OK, I was able to reproduce this when generating the ACK EC2 controller. I'm still not sure why it is happening, but certain shapes in the API model -- TagSpecification, VolumeAttachment, and others -- are not getting generated properly and are causing `deepcopy-gen` to fail. I will keep looking into this.

**Comment by haarchri** [comment]
Thanks for information - can we do something on our side ? ;)

**Comment by RedbackThomson** [comment]
The reason why `VolumeAttachment` is not being generated is because it is a member of the `Volume` type. We have an explicit check in the code that any structures that are "Payloads" will not have their types, or their members, generated. We define a "Payload" as any type that is used as the input or the output from a `Create` operation. The [`CreateVolume`](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.CreateVolume) operation uses `Volume` as its output type and therefore our generator will ignore it, whereas we would hope it would continue to https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/model/model.go#L314-L322.

I have a PR open that checks to see whether any fields in the Spec are using the payload's structure directly (which in this case would be `Volume`). This same principle could be expanded to allow any fields in Status, and for any field whose members are required (`Volume`'s member `VolumeAttachment`).

**Comment by RedbackThomson** [comment]
Slight correction on the post above:
We actually iterate through all API shapes - which would mean we iterate over both `Volume` and `VolumeAttachment`. While what I typed is true for `Volume`, the loop would eventually reach `VolumeAttachment`. Funnily enough, we are also registering `VolumeAttachment` is a "payload" because it is the output of the [`DetachVolume`](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DetachVolume) operation. 

Regardless, the fix I mentioned above will still apply - however we don't need to worry about the "and for any field whose members are required (Volume's member VolumeAttachment)." part. We can re-use the changes I made in [this PR](https://github.com/aws-controllers-k8s/code-generator/pull/139/files), replacing `spec` with `status`.
